qiskit_experiments/visualization/plotters/base_plotter.py

```diff
@@ -0,0 +1,453 @@
+        ]
+        for unknown_data_key in unknown_data_keys:
+            warnings.warn(
+                f"{self.__class__.__name__} encountered an unknown data-key {unknown_data_key}. It may "
+                "not be used by the plotter class."
+            )
+
+        self._figure_data.update(**data_kwargs)
+
+    def clear_figure_data(self):
+        """Clears figure data."""
+        self._figure_data = {}
+
+    def data_exists_for(self, series_name: str, data_keys: Union[str, List[str]]) -> bool:
+        """Returns whether the given data-keys exist for the given series.
+
+        Args:
+            series_name: The name of the given series.
+            data_keys: The data-keys to be checked.
+
+        Returns:
+            bool: True if all data-keys have values assigned for the given series. False if at least one
+                does not have a value assigned.
+        """
+        if not isinstance(data_keys, list):
+            data_keys = [data_keys]
+
+        # Handle non-existent series name
+        if series_name not in self._series_data:
+            return False
+
+        return all(key in self._series_data[series_name] for key in data_keys)
+
+    @abstractmethod
+    def _plot_figure(self):
+        """Generates a figure using :attr:`drawer` and :meth:`data`.
+
+        Sub-classes must override this function to plot data using the drawer. This function is called by
+        :meth:`figure` when :attr:`drawer` can be used to draw on the canvas.
+        """
+
+    def figure(self) -> Any:
+        """Generates and returns a figure for the already provided series and figure data.
+
+        :meth:`figure` calls :meth:`_plot_figure`, which is overridden by sub-classes. Before and after
+        calling :meth:`_plot_figure`; :func:`_initialize_drawer`, :func:`initialize_canvas` and
+        :func:`format_canvas` are called on the drawer respectively.
+
+        Returns:
+            Any: A figure generated by :attr:`drawer`, of the same type as ``drawer.figure``.
+        """
+        # Initialize drawer, to copy axis, subplots, and plot-options across.
+        self._initialize_drawer()
+
+        # Initialize canvas, which creates subplots, assigns axis labels, etc.
+        self.drawer.initialize_canvas()
+
+        # Plot figure for given subclass. This is the core of BasePlotter subclasses.
+        self._plot_figure()
+
+        # Final formatting of canvas, which sets axis limits etc.
+        self.drawer.format_canvas()
+
+        # Return whatever figure is created by the drawer.
+        return self.drawer.figure
+
+    @classmethod
+    @abstractmethod
+    def expected_series_data_keys(cls) -> List[str]:
+        """Returns the expected series data-keys supported by this plotter."""
+
+    @classmethod
+    @abstractmethod
+    def expected_figure_data_keys(cls) -> List[str]:
+        """Returns the expected figures data-keys supported by this plotter."""
+
+    @property
+    def options(self) -> Options:
```

Comment:
I completely understand that. @eggerdj suggested `drawer_options` or `options_for_drawer`, which are also good. In hindsight, it would be good for users to be oblivious of the drawer unless otherwise necessary. If that's the case, maybe the options should be renamed to `options` and `figure_options`; where `figure_options` are options/parameters for the figure itself (like axis labels etc.) whereas `options` are parameters that control the behaviour of the plotter or drawer. I've made this change in 956696a, but we can continue discussing it here.